tools/clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -762,7 +763,18 @@ void clang::EmitBackendOutput(DiagnosticsEngine &Diags,
                               raw_pwrite_stream *OS) {
   EmitAssemblyHelper AsmHelper(Diags, CGOpts, TOpts, LOpts, M);
 
-  AsmHelper.EmitAssembly(Action, OS);
+  // Catch any fatal errors during optimization passes here
+  // so that future passes can be skipped.
+  try {
+    AsmHelper.EmitAssembly(Action, OS);
+  } catch (const ::hlsl::Exception &hlslException) {
+    const char *msg = hlslException.what();
+    const std::string msgStr(msg);
```

Comment:
I believe a `llvm::StringRef` will suffice here instead of the `std::string`, avoiding a potential string copy.